flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/TestingPhysicalSlotProvider.java

```diff
@@ -0,0 +1,166 @@
+                        CompletableFuture.completedFuture(
+                                new TestingPhysicalSlot(resourceProfile, new AllocationID())));
+    }
+
+    public static TestingPhysicalSlotProvider createWithoutImmediatePhysicalSlotCreation() {
+        return create((ignored) -> new CompletableFuture<>());
+    }
+
+    public static TestingPhysicalSlotProvider createWithFailingPhysicalSlotCreation(Throwable t) {
+        return create((ignored) -> FutureUtils.completedExceptionally(t));
+    }
+
+    public static TestingPhysicalSlotProvider createWithLimitedAmountOfPhysicalSlots(
+            int slotCount) {
+        return createWithLimitedAmountOfPhysicalSlots(
+                slotCount, new SimpleAckingTaskManagerGateway());
+    }
+
+    public static TestingPhysicalSlotProvider createWithLimitedAmountOfPhysicalSlots(
+            int slotCount, TaskManagerGateway taskManagerGateway) {
+        AtomicInteger availableSlotCount = new AtomicInteger(slotCount);
+        return create(
+                (resourceProfile) -> {
+                    int count = availableSlotCount.getAndDecrement();
+                    if (count > 0) {
+                        return CompletableFuture.completedFuture(
+                                TestingPhysicalSlot.builder()
+                                        .withResourceProfile(resourceProfile)
+                                        .withTaskManagerGateway(taskManagerGateway)
+                                        .build());
+                    } else {
+                        return FutureUtils.completedExceptionally(
+                                new NoResourceAvailableException(
+                                        String.format(
+                                                "The limit of %d provided slots was reached. No available slots can be provided.",
+                                                slotCount)));
+                    }
+                });
+    }
+
+    private TestingPhysicalSlotProvider(
+            Function<ResourceProfile, CompletableFuture<TestingPhysicalSlot>> physicalSlotCreator) {
+        this.physicalSlotCreator = physicalSlotCreator;
+
+        this.requests = new HashMap<>();
+        this.responses = new HashMap<>();
+        this.cancellations = new HashMap<>();
+    }
+
+    @Override
+    public CompletableFuture<PhysicalSlotRequest.Result> allocatePhysicalSlot(
+            PhysicalSlotRequest physicalSlotRequest) {
+        SlotRequestId slotRequestId = physicalSlotRequest.getSlotRequestId();
+        requests.put(slotRequestId, physicalSlotRequest);
+        CompletableFuture<TestingPhysicalSlot> resultFuture;
+        try {
+            resultFuture =
+                    physicalSlotCreator.apply(
+                            physicalSlotRequest.getSlotProfile().getPhysicalSlotResourceProfile());
+        } catch (Throwable t) {
+            resultFuture = FutureUtils.completedExceptionally(t);
+        }
```

Comment:
Are we still needing this catch block?

Reply:
Ah thanks for removing it. That one I missed accidentally... 👍